sql: adjust "update" portion of UPDATEs and UPSERTs

yuzefovich opened the pull request:
**execbuilder: expand output in secondary_index_column_families**

This commit adjusts `secondary_index_column_families` file to show
all KV operations from the trace. This is helpful in understanding of
the behavior of some changes related to buffered writes.

It also fixes the filter used in `TestPrimaryKeyChangeKVOps` that was
broken in 5a97a11 (i.e. not all operations were shown).

**sql: adjust "update" portion of UPDATEs and UPSERTs**

This commit finalizes the adjustments that we need for UPDATEs and
UPSERTs with regards to the buffered writes. Whenever we have a primary
key change, we simply DELETE the full old row and then INSERT the full
new row, and this is already handled correctly. (A TODO for an
optimization of this code-path in some cases has been added.)

Whenever we don't have the primary key change, we want the following
behavior achieved in this commit:
- locking `Put`s on the primary index (no need for `CPut` since the
primary key isn't changing)
- `CPut`s on unique secondary indexes, unless there is no key change in
which case we need locking `Put`s
- non-locking `Put`s on non-unique secondary indexes.

Note that this commit also adjusts the INSERT code path to use
non-locking `Put`s on non-unique secondary indexes. In turn, this
triggered an adjustment to the BatchEncoder used by COPY which imitates
the INSERT row-by-row path.

As an optimization we can elide the lock acquisitions if we already
locked the corresponding index during the initial scan. This commit
implements this logic with this optimization.

One scenario I'm concerned about is when we have at least 3 column
families - it seems possible that during the initial scan we only lock
_some_ column families by issuing `Get`s as opposed to a `Scan` over the
whole row. In such a case we might not be able to elide the lock
acquisitions even for indexes that were locked during the initial scan.
I left a TODO to think about this scenario and will address separately
(either by proving that we don't use "family-specific spans", meaning we
always will scan and locks full rows, or by adjusting the lock eliding
logic).

Informs: #139106.
Fixes: #139160.

**sql: disable lock eliding optimization for multiple column families**

This commit disables the "lock eliding" optimization (that we do with
buffered writes when we performed the locking reads during the initial
scan of the mutation) when the table has multiple column families. In such
a scenario it is possible that the locking read will be handled by
family-specific locking Gets meaning that we don't actually lock the
full rows. The contract of the `Updater` and the `Deleter` is such that
the caller guarantees all possible old keys are locked, and in this
scenario it might not be the case, so we choose to disable the
optimization.

Note that in practice column families tend to not be used often (at
least according to our telemetry as of a few years ago), so this
shouldn't affect most users. Additionally, we have an idea of how to
push the lock eliding optimization into the KV client (since the
TxnCoordSender knows the full set of already locked keys, it'll be much
easier to handle the optimization there), so spending time on being more
precise when the optimization is disabled at the SQL level doesn't seem
worth it.

Fixes: #139106.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143247)
<!-- Reviewable:end -->

yuzefovich committed:
execbuilder: expand output in secondary_index_column_families

This commit adjusts `secondary_index_column_families` file to show
all KV operations from the trace. This is helpful in understanding of
the behavior of some changes related to buffered writes.

It also fixes the filter used in `TestPrimaryKeyChangeKVOps` that was
broken in 5a97a11337dbd3874bea2342a727d8b4500ac224 (i.e. not all
operations were shown).

Release note: None

yuzefovich committed:
sql: adjust "update" portion of UPDATEs and UPSERTs

This commit finalizes the adjustments that we need for UPDATEs and
UPSERTs with regards to the buffered writes. Whenever we have a primary
key change, we simply DELETE the full old row and then INSERT the full
new row, and this is already handled correctly. (A TODO for an
optimization of this code-path in some cases has been added.)

Whenever we don't have the primary key change, we want the following
behavior achieved in this commit:
- locking `Put`s on the primary index (no need for `CPut` since the
primary key isn't changing)
- `CPut`s on unique secondary indexes, unless there is no key change in
which case we need locking `Put`s
- non-locking `Put`s on non-unique secondary indexes.

Note that this commit also adjusts the INSERT code path to use
non-locking `Put`s on non-unique secondary indexes. In turn, this
triggered an adjustment to the BatchEncoder used by COPY which imitates
the INSERT row-by-row path.

As an optimization we can elide the lock acquisitions if we already
locked the corresponding index during the initial scan. This commit
implements this logic with this optimization.

One scenario I'm concerned about is when we have at least 3 column
families - it seems possible that during the initial scan we only lock
_some_ column families by issuing `Get`s as opposed to a `Scan` over the
whole row. In such a case we might not be able to elide the lock
acquisitions even for indexes that were locked during the initial scan.
I left a TODO to think about this scenario and will address separately
(either by proving that we don't use "family-specific spans", meaning we
always will scan and lock full rows, or by adjusting the lock eliding
logic).

Release note: None

michae2 reviewed (commented):
I'm still reading through tests, but so far this looks good.

> Reviewed 35 of 35 files at r5, 24 of 35 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143247#-:-OM8dj087mzmauxWA7yH:b-1xno2v)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @stevendanna)

___
*[`-- commits` line 48 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/143247#-OM89p3l1bO42W8TAyzW:-OM89p3l1bO42W8TAyzX:bywycg5):*
This is a great point. Looking at this example:

```sql
CREATE TABLE abcde (a INT PRIMARY KEY, b INT, c INT, d INT, e INT, INDEX (e), FAMILY (a), FAMILY (b), FAMILY (c), FAMILY (d), FAMILY (e));
INSERT INTO abcde VALUES (4, 5, 6, 7, 8);
\set auto_trace=kv
EXPLAIN ANALYZE (VERBOSE) UPDATE abcde SET c = 9 WHERE a = 4;
EXPLAIN ANALYZE (VERBOSE) DELETE FROM abcde WHERE a = 4;
```

During the UPDATE the modified column (c) is also one of the fetch columns (a, c), even though the modification doesn't use the old value. So that does get the lock on column family 2 during the initial scan:

```
Scan /Tenant/3/Table/104/1/4/0, /Tenant/3/Table/104/1/4/2/1 lock Exclusive (Block, Unreplicated)
```

But during the DELETE, only columns (a, e) are fetched, so none of the other column families are locked during the initial scan. And yet none of the Dels are locked:

```
Scan /Tenant/3/Table/104/1/4/0, /Tenant/3/Table/104/1/4/4/1 lock Exclusive (Block, Unreplicated)
...
Del /Tenant/3/Table/104/2/8/4/0
Del /Tenant/3/Table/104/1/4/0
Del /Tenant/3/Table/104/1/4/1/1
Del /Tenant/3/Table/104/1/4/2/1
Del /Tenant/3/Table/104/1/4/3/1
Del /Tenant/3/Table/104/1/4/4/1
```

So that does look like an issue.

(Also, we won't lock column families with all NULL values during the initial scan, since the family doesn't exist as a KV. But maybe that's ok as long as we're always using a CPut to write the new KV when setting one of those columns to a non-NULL value? For example, we decided https://github.com/cockroachdb/cockroach/issues/116836 is probably ok.) 
___
*[`pkg/sql/colenc/inverted.go` line 97 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/143247#-OM8VRGoAiaAyeieiAT5:-OM8VRGoAiaAyeieiAT6:bkv6n8o) ([raw file](https://github.com/cockroachdb/cockroach/blob/218377b4f0816f8c9f09d527a695f10153386697/pkg/sql/colenc/inverted.go#L97)):*
> ```Go
> 	var kvValue roachpb.Value
> 	kvValue.SetBytes(value)
> 	if ind.IsUnique() {
> ```

This doesn't hurt anything, but can inverted indexes be unique? :thinking: I didn't think so.
___
*[`pkg/sql/row/updater.go` line 397 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/143247#-OM8XDYC7kgVKwlHqx61:-OM8XDYC7kgVKwlHqx62:b-4l8xcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/218377b4f0816f8c9f09d527a695f10153386697/pkg/sql/row/updater.go#L397)):*
> ```Go
> 		// secondary indexes when the row doesn't contain NULL values in the
> 		// indexed columns. In such a scenario, the key in the unique secondary
> 		// index doesn't change, so rather than performing a Del followed by a
> ```

This is a great realization!

micro-nit: maybe I would say "doesn't _necessarily_ change" here, since the unique secondary index key could change if it is also assigned to in the update statement


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143247)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @michae2, and @stevendanna)

___
*[`-- commits` line 48 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/143247#-OM89p3l1bO42W8TAyzW:-OMDtOV16GZwvYEBpBmj:b-8a5tvo):*
Ack, thanks for looking into this. I'll track this issue as the last item in #139106 and will address in a separate PR.

> we won't lock column families with all NULL values during the initial scan, since the family doesn't exist as a KV. But maybe that's ok as long as we're always using a CPut to write the new KV when setting one of those columns to a non-NULL value?

Good point. The only scenario where we won't do a locking write is when we don't have the PK change, we locked the primary index during the initial scan, and we need to insert a new KV for the previously skipped only NULL column family that is no longer fully NULL. In such case, we ask for a locking Get (but we don't get the lock since we don't set LockNonExisting flag) and we'll use a non-locking Put, so I left a TODO for now. This seems like an edge case though, so perhaps we won't do anything about it (similar to #116836).
___
*[`pkg/sql/colenc/inverted.go` line 97 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/143247#-OM8VRGoAiaAyeieiAT5:-OMDsB-N9IfI9CaJlQiD:bn3qcpq) ([raw file](https://github.com/cockroachdb/cockroach/blob/218377b4f0816f8c9f09d527a695f10153386697/pkg/sql/colenc/inverted.go#L97)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

This doesn't hurt anything, but can inverted indexes be unique? :thinking: I didn't think so.
</blockquote></details>

No, I don't think either, I included this only for completeness (we already have a similarly redundant check in `encodeInvertedSecondaryIndex` above).
___
*[`pkg/sql/row/updater.go` line 397 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/143247#-OM8XDYC7kgVKwlHqx61:-OMDsmJ-03-l7LiU5Khp:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/218377b4f0816f8c9f09d527a695f10153386697/pkg/sql/row/updater.go#L397)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

This is a great realization!

micro-nit: maybe I would say "doesn't _necessarily_ change" here, since the unique secondary index key could change if it is also assigned to in the update statement
</blockquote></details>

Good point, done.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
sql: disable lock eliding optimization for multiple column families

This commit disables the "lock eliding" optimization (that we do with
buffered writes when we performed the locking reads during the initial
scan of the mutation) when the table has multiple column families. In such
a scenario it is possible that the locking read will be handled by
family-specific locking Gets meaning that we don't actually lock the
full rows. The contract of the `Updater` and the `Deleter` is such that
the caller guarantees all possible old keys are locked, and in this
scenario it might not be the case, so we choose to disable the
optimization.

Note that in practice column families tend to not be used often (at
least according to our telemetry as of a few years ago), so this
shouldn't affect most users. Additionally, we have an idea of how to
push the lock eliding optimization into the KV client (since the
TxnCoordSender knows the full set of already locked keys, it'll be much
easier to handle the optimization there), so spending time on being more
precise when the optimization is disabled at the SQL level doesn't seem
worth it.

Release note: None

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143247)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @michae2, and @stevendanna)

___
*[`-- commits` line 48 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/143247#-OM89p3l1bO42W8TAyzW:-OMJF85i9rdhUptpHXLW:b-gre613):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Ack, thanks for looking into this. I'll track this issue as the last item in #139106 and will address in a separate PR.

> we won't lock column families with all NULL values during the initial scan, since the family doesn't exist as a KV. But maybe that's ok as long as we're always using a CPut to write the new KV when setting one of those columns to a non-NULL value?

Good point. The only scenario where we won't do a locking write is when we don't have the PK change, we locked the primary index during the initial scan, and we need to insert a new KV for the previously skipped only NULL column family that is no longer fully NULL. In such case, we ask for a locking Get (but we don't get the lock since we don't set LockNonExisting flag) and we'll use a non-locking Put, so I left a TODO for now. This seems like an edge case though, so perhaps we won't do anything about it (similar to #116836).
</blockquote></details>

Actually, given that we're having this discussion, it seems easier to include the commit handling multiple column families into this PR too. I decided to simply disable the lock eliding optimization on tables with 3+ column families due to:
- it's actually uncommon to have multiple column families (I checked telemetry a few years ago for this)
- we have an idea of how to push the lock eliding optimization into the KV client.

I'll add a session variable to control this behavior though so that we can enable the optimization if we really need it, but I feel like spending time trying to figure out the exact cases when we will or will not lock all the keys being modified is not warranted.


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice job!

> Reviewed 21 of 21 files at r8, 5 of 6 files at r9, 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143247#-:-OMJL5r_6iJzBQ5uhpKP:b-46j5c8)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani and @stevendanna)

___
*[`-- commits` line 45 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/143247#-OMJh591AYbhr_Ado4uv:-OMJh0pE1ujlwIW5wxs_:b-2sr8wb):*
I'm glad we're changing writes to non-unique secondary indexes to use Put, but it also seems like one of the biggest risks of this work in 25.2, since it will affect most insert and update statements even with buffered writes disabled.

Is the plan for #139164 to enable / disable this independently of buffered writes being enabled / disabled?
___
*[`-- commits` line 83 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/143247#-OMJt_XeCALi3e06kyxD:-OMJt_XeCALi3e06kyxE:btm4ox9):*
Ah nice! That would be much simpler (assuming we already have to look up the key in the buffer).
___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 50 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/143247#-OMJvHvD888RBDy2ulWi:-OMJvHvEA4HNLzd0Uwr0:b-y0ntwq) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d13de31401176177eb148436ed6b95be84c71ae/pkg/sql/opt/exec/execbuilder/mutation.go#L50)):*
> ```Go
> 			b.forceForUpdateLocking = 0
> 		}()
> 		if b.mem.Metadata().Table(toLock).FamilyCount() >= 3 {
> ```

Is this >= 3 rather than >= 2 because SplitRowKeyIntoFamilySpans merges adjacent spans? I wonder if there's a way to tie this check to that behavior? Maybe a comment at least?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143247#-:-OMK6UzbBAymLfprmvTS:b-oduykh)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani, @michae2, and @stevendanna)

___
*[`-- commits` line 45 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/143247#-OMJh591AYbhr_Ado4uv:-OMK3F9I3JHNy90xAHQo:b3tuzjr):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I'm glad we're changing writes to non-unique secondary indexes to use Put, but it also seems like one of the biggest risks of this work in 25.2, since it will affect most insert and update statements even with buffered writes disabled.

Is the plan for #139164 to enable / disable this independently of buffered writes being enabled / disabled?
</blockquote></details>

I agree about the risk here. Correct, the plan for the knob is to force usage of CPuts instead of Puts for non-unique indexes regardless of whether the buffered writes are enabled. I'm thinking of adding two other knobs that will only be applicable for buffered writes.
___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 50 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/143247#-OMJvHvD888RBDy2ulWi:-OMK5YGTC7K1o-sQL_jm:b-qwiz9r) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d13de31401176177eb148436ed6b95be84c71ae/pkg/sql/opt/exec/execbuilder/mutation.go#L50)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Is this >= 3 rather than >= 2 because SplitRowKeyIntoFamilySpans merges adjacent spans? I wonder if there's a way to tie this check to that behavior? Maybe a comment at least?
</blockquote></details>

Good question. I was wrong - 2 column families can still be problematic (like we could issue a locking Get just for 0th family without locking 1st), so we need to disable the optimization for 2 families as well. I initially was thinking that we needed to have at least 3 families so that the read of the row would be `Get(0th), Get(2nd)`, but we can still do `Get(0th)` with just 2 families total. Updated the code and the commit message, thanks for the catch!


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Oops, need to update one more test.

bors r-

yuzefovich commented:
bors r+
